h2/src/main/org/h2/result/ResultTempTable.java

```diff
@@ -89,25 +91,40 @@ private ResultTempTable(ResultTempTable parent) {
 
     private void createIndex() {
         IndexColumn[] indexCols = null;
-        // If we need to do distinct, the distinct columns may not match the
-        // sort columns. So we need to disregard the sort. Not ideal.
-        if (sort != null && !distinct) {
+        if (sort != null) {
             int[] colIndex = sort.getQueryColumnIndexes();
-            indexCols = new IndexColumn[colIndex.length];
-            for (int i = 0; i < colIndex.length; i++) {
-                IndexColumn indexColumn = new IndexColumn();
-                indexColumn.column = table.getColumn(colIndex[i]);
+            int len = colIndex.length;
+            int totalLen;
+            BitSet used;
+            if (distinct) {
+                totalLen = columnCount;
+                used = new BitSet();
+            } else {
+                totalLen = len;
+                used = null;
+            }
+            indexCols = new IndexColumn[totalLen];
+            for (int i = 0; i < len; i++) {
+                int idx = colIndex[i];
+                if (used != null) {
+                    used.set(idx);
+                }
+                IndexColumn indexColumn = createIndexColumn(idx);
                 indexColumn.sortType = sort.getSortTypes()[i];
-                indexColumn.columnName = COLUMN_NAME + i;
                 indexCols[i] = indexColumn;
             }
+            if (used != null) {
+                int idx = 0;
+                for (int i = len; i < totalLen; i++) {
+                    idx = used.nextClearBit(idx);
+                    indexCols[i] = createIndexColumn(idx);
+                    idx++;
+                }
+            }
```

Comment:
I split this into three blocks instead of two.

`SELECT DISTINCT a,b FROM test ORDER BY b` should be OK, similar situation is tested. `ResultTempTable` now creates an index with `b, a` ordering.

Before my changes `a, b` index was used incorrectly.